pallets/xcm-core-buyer/src/lib.rs

```diff
@@ -261,6 +379,62 @@ pub mod pallet {
 
             Ok(())
         }
+
+        #[pallet::call_index(4)]
+        #[pallet::weight(T::WeightInfo::set_relay_chain())] //TODO: Proper benchmark
+        pub fn query_response(
+            origin: OriginFor<T>,
+            query_id: QueryId,
+            response: Response,
+        ) -> DispatchResult {
+            let _responder = ensure_response(<T as Config>::RuntimeOrigin::from(origin))?;
+
+            let mut query_id_to_para_id = QueryIdToParaId::<T>::get();
+            let maybe_para_id = query_id_to_para_id.get(&query_id);
+
+            let para_id = if let Some(para_id) = maybe_para_id {
+                *para_id
+            } else {
+                // Most probably entry was expired or removed in some other way. Let's return early.
+                return Ok(());
+            };
+
+            let mut in_flight_orders = InFlightOrders::<T>::get();
+            let order = in_flight_orders.get(&para_id).expect("If the QueryId->ParaId mapping exists \
+                    then an entry must be present in the InFlightOrders, if not we should crash as that \
+                    indicates storage inconsistency; qed");
```

Comment:
How is `query_response` executed? If it is on `on_initialize`, or some other mandatory mechanism, we should never crash, even if the storage is inconsistent, because that will stall the chain.

A conservative apporach would be to ignore any messages when the state is inconsistent, which can be easily achieved by retuning an error here.

```suggestion
            let order = in_flight_orders.get(&para_id).ok_or(Error::<T>::InFlightOrdersInconsistentState)?;
```

Reply:
it's part of the XCM message which now are executed on_initialize. Yes we should never crash, but rather error

Reply:
Also, why do we need a response? Wasn't xcm "fire and forget"?

Reply:
I am not sure here. Reason I chose crash is because if there is no corresponding entry, it could indicate storage inconsistency. 

In that situation, lets the chain work as usual would mean that we are operating on inconsistent state. Instead shouldn't we try to fix it before we allow any child blocks to be executed?

Reply:
I get what you mean, but since this is executed in on_initialize, we cannot fix it, because any extrinsic we send will be executed after the panic, so it won't be executed. You can use this substrate macro to warn in production but panic in tests, this way hopefully we will detect it if anything goes wrong:

https://paritytech.github.io/polkadot-sdk/master/frame_support/macro.defensive.html

Reply:
Hmm. If I only have one data structure, I think this problem is eliminated.